Add fedora-integration: Support `override replace https://bodhi/...`

cgwalters opened the pull request:
This adds support for e.g.:

```
$ rpm-ostree override replace https://bodhi.fedoraproject.org/updates/FEDORA-2020-2908628031
```

This will find the Koji builds from the listed update, download
all the RPMs (that aren't debuginfo) and pass them for overrides
in the same way we support `override replace http://somewebserver/foo.rpm`
now.

We also support directly linking a Koji build:
```
$ rpm-ostree override replace https://koji.fedoraproject.org/koji/buildinfo?buildID=1625029
```

There are some parts of this code that are *very* hacky, most
notably how we parse the Apache directory listings from kojipkgs
rather than speak XML-RPC (or drag in the `koji` package and
its large set of Python deps).

Bodhi at least has a modern HTTP+JSON API.

Also notably this still downloads *all* the other RPMs even
ones that aren't installed.  Handling that truly correctly
would require moving this logic to the daemon and core.

All of this functionality is keyed off a `cfg(feature = "fedora-integration")`
that is detected by a Rust `build.rs` which parses the build environment's
`/etc/os-release` for now.

cgwalters commented:
(Draft since I based this on https://github.com/coreos/rpm-ostree/pull/2416 because TBH I'm trying to limit the amount of time I spend resolving conflicts between my own PRs)

cgwalters commented:
Though it would probably be relatively straightforward to write (and hopefully deploy) a proxy service that queries Koji via XML-RPC and returns JSON that we could parse instead.

cgwalters commented:
OK rebased :surfer: and should be ready for review!

cgwalters commented on build.rs:
```diff
@@ -0,0 +1,21 @@
+    Ok(())
+}
+
+fn main() -> Result<()> {
```
Note this introduction of `build.rs` is something we'll use in the future for sure too.

cgwalters commented:
Filed https://github.com/fedora-infra/bodhi/issues/4175
But I don't think we need to block on this - scraping HTML is ugly, but if this breaks it's not "critical path" functionality, and I would bet money on nothing interesting changing in Koji in 2021 (or 2022, or...).

jlebon commented on rust/src/fedora_integration.rs:
```diff
@@ -0,0 +1,239 @@
+        if !l.starts_with("<img") {
+            return Ok(r);
+        }
+        for v in l.split('"') {
+            if v.ends_with(".rpm") {
+                r.push(v.to_string())
+            }
+        }
+        Ok(r)
+    })
+}
+
+/// Scrape the Koji HTML and find the buildid (again until Koji gains some sort
+/// of modern JSON API)
+fn get_nvr_from_koji_html(buildid: &str, html: impl Read) -> Result<String> {
+    let r = std::io::BufReader::new(html);
+    let self_anchor = format!("<a href=\"buildinfo?{}\">", buildid);
+    for line in r.lines() {
+        let line = line?;
+        let idx = if let Some(idx) = line.find(&self_anchor) {
+            idx + self_anchor.len()
+        } else {
+            continue;
+        };
+        return Ok(line[idx..].split('<').next().expect("split").to_string());
+    }
+    Err(anyhow::anyhow!("Failed to find NVR from Koji HTML"))
+}
+
+fn get_buildid_from_koji_url(url: &str) -> Result<&str> {
+    let id = url.rsplit('?').next().expect("split");
+    if !id.starts_with("buildID=") {
+        anyhow::bail!("Failed to parse Koji buildid from URL {}", url);
+    }
+    Ok(id)
+}
+
+fn is_debug_rpm(rpm: &str) -> bool {
+    rpm.contains("-debuginfo-") || rpm.contains("-debugsource-")
+}
```
Also, seems like this is not accounting for `noarch` subpackages.

cgwalters replied:
> This kinda hurts my eyes and soul to review. :)

Oh yeah, I felt the same when writing it :smile: 

>  to just use an XML parser.

HTML isn't XML - [that attempt failed](https://en.wikipedia.org/wiki/XHTML).  Now there is [a Rust HTML5 parser](https://github.com/servo/html5ever) but the depenency chain for that is large and nothing we do there is going to be seriously robust to markup changes anyways.

> Should also make it easier to extract the RPM links from the Koji page itself and avoid having to also fetch and parse the Apache listing.

Hmm...hard to say for sure whether it's more likely he Koji HTML changes or the Apache directory listing does.  Both are pretty fixed in format.

> Also, seems like this is not accounting for noarch subpackages.

Ah yes, you're right; will look at fixing that.

travier replied:
Should this ignore `-devel` packages too?

cgwalters replied:
Re `-devel`...yeah perhaps, but they're much smaller, and also at least for Silverblue we ship some kernel devel packages by default, so it's more likely people have them.

jlebon reviewed (commented):
Cool idea overall!

> Also notably this still downloads all the other RPMs even
ones that aren't installed.

Yeah, that can be really costly depending on the package.

>  Handling that truly correctly
would require moving this logic to the daemon and core.

Meh, I'd vote for continuing to do this client-side to be safe and just query the rpmdb from the pending deployment via `rpm_ostree_db_query_all` -- we do something similar to print pkgdiffs client-side in `rpmostree_print_treepkg_diff_from_sysroot_path`. (Then obviously, parse the NEVRA of each RPM, and only add to the list to download those whose names are in the rpmdb.)

cgwalters commented:
OK moved the prep code for this to https://github.com/coreos/rpm-ostree/pull/2429

Maybe we can give this a day and see what the Bodhi maintainers think; if they agree to implement it then we can probably just wait for that.

cgwalters commented on rust/src/client.rs:
```diff
@@ -0,0 +1,20 @@
+    arg.starts_with("https://") || arg.starts_with("http://")
+}
+
+pub(crate) fn client_handle_fd_argument(arg: &str, arch: &str) -> Result<Vec<i32>> {
```
Looked at that but I think it's even cleaner to move the "ends in .rpm" logic into Rust too, so it says `client_handle_fd_argument()`.  See https://github.com/coreos/rpm-ostree/pull/2429

cgwalters commented:
OK a problem with handling `noarch` is not every package has them...and handling that correctly is more ugliness.

cgwalters commented:
OK I wrote
https://github.com/cgwalters/koji-sane-json-api
and deployed it, see e.g.:
https://kojiproxy-coreos.svc.ci.openshift.org/buildinfo/NetworkManager-1.26.4-1.fc33

And then rewrote this PR to use it and it's obviously *much* saner.

cgwalters commented:
We also now correctly pull in `noarch` packages if they exist.

jlebon commented on rust/src/fedora_integration.rs:
```diff
@@ -0,0 +1,151 @@
+
+    pub(crate) fn get_buildid_from_url(url: &str) -> Result<&str> {
+        let id = url.rsplit('?').next().expect("split");
+        if !id.starts_with("buildID=") {
```
Another really cool future addition is support for taskID URLs. Then it'd be super easy to test out scratch builds too.

jlebon reviewed (approved):
Could we exercising this in the `vmcheck` suite? There's the `test-override-kernel.sh` where we download the kernel which would be a good candidate (e.g. downloading that kernel from the Bodhi URL; that way it implicitly tests most of the Koji bits too).

Hmm actually, that's currently the only test which exercises the URL-to-RPM path, so we'd lose that coverage. Though maybe we can just add a URL to some other pkg in that same invocation, e.g. to `strace` or something. That way we don't increase the number of places we have to manually bump when moving to a new releasever.

cgwalters commented:
OK yep thanks for calling out the testing, it was actually really easy to write this and revealed I did break the koji URL path in the last refactoring.  Debugging that took excessively log because the proxy service wasn't logging errors (now also fixed).

cgwalters commented:
/retest

cgwalters commented:
/retest

jlebon commented:
/lgtm

cgwalters commented:
The `layering-basic-1` failure there is odd.  I don't see how it could be related to this PR, and it passes locally.  But I don't see how it could be a flake either.  Seems like somehow we ended up adding both `foo-1.0` and `foo-ext` to the `packages` when the former should just be a dependency.

cgwalters commented:
I added another small commit here to propagate it as a feature and rebased :surfer: on master to pick up https://github.com/coreos/rpm-ostree/pull/2436 merged!

jlebon commented on rust/src/utils.rs:
```diff
@@ -294,3 +294,10 @@ pub(crate) fn varsubstitute(s: &str, subs: &Vec<crate::ffi::StringMapping>) -> R
+pub(crate) fn get_features() -> Vec<String> {
+    let mut r = Vec::new();
+    #[cfg(feature = "fedora-integration")]
+    r.push("fedora-infrastructure".to_string());
```
Did you mean `fedora-integration` here?

cgwalters replied:
Yeah, better to be consistent; fixed!

jlebon commented:
/lgtm
